### ISSUE TITLE
Support for regex rewrite in contour

### ISSUE THREAD

**Issue by gsadasiv** [projectcontour/contour#4241]
Looking for a capability to rewrite path in HTTP proxy.

e.g. /f1/<f1-id>/f2/<f2-id>/f3/<f3-id>/…  -> /f1/<f1-id>/f3/<f3-id>/…

Envoy seems to have this capability through regex_rewrite. https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-r[…]v3-routeaction-regex-rewrite

Can this be exposed in Contour?

**Comment by youngnick** [comment]
Contour already supports rewriting the request path in HTTPProxy, see https://projectcontour.io/docs/v1.19.1/config/request-rewriting/ for the current-stable docs on this.

Notably, however, it does *not* support regexes for either field - they must be exact matches.

There are two reasons for this:
- Firstly, it's difficult to validate that the regex will be accepted by Envoy and because of #1176, if we send Envoy an update that it can't accept, it may stop Envoy accepting config changes from Contour until the change is reverted. Basically, we can't validate that the regex will be accepted, and if a user put in an invalid regex, Envoy would silently stop accepting updates until the invalid regex was fixed. And you'd have no way to know, because Contour currently can't easily tell you that it was that config that produced the problem.
- Secondly, even if we fix those problems, Envoy keeps track of a metric called "regex size", which is a measure of the computational complexity of the regex. If the regex exceeds that size, it may also be rejected. #2043 has the details.

The tl;dr here is that:
- you can already do some path replacement
- path replacement must use literal strings, not a regex, unfortunately.

Hopefully this helps @gsadasiv.

**Comment by skriss** [comment]
From @pavelnikolov in #4364:

> **Please describe the problem you have**
> 
> **tl;dr** I am trying to rewrite my path from `/prometheus/account/*/api/v1/query` to `/admin/prometheus/api/v1/query`.
> 
> I am using Contour as ingress controller in my Kubernetes cluster. I have a multi-tenanted Prometheus and I want to allow my users to access three paths `/api/v1/query`, `/api/v1/query_range` and `/federate`. I am using [PromAuthProxy](https://github.com/cherti/PromAuthProxy) to restrict the returned metrics if the user is authorized to access the metrics for the given account ID.
> 
> This is my `HttpProxy` configuration:
> 
> ```yaml
> # inside the root HttpProxy
>   ...
>   includes:
>     ...
>     - name: promauthproxy
>       namespace: customer-monitoring
>       conditions:
>         - prefix: /prometheus/account
>    ...
> ```
> 
> this is the `promauthproxy` definition:
> 
> ```yaml
> apiVersion: projectcontour.io/v1
> kind: HTTPProxy
> metadata:
>   name: promauthproxy
>   namespace: customer-monitoring
> spec:
>   routes:
>     - pathRewritePolicy:
>         replacePrefix:
>           - prefix: /prometheus/account/*/federate
>             replacement: /admin/prometheus/federate
>           - prefix: /prometheus/account/*/api/v1/query
>             replacement: /admin/prometheus/api/v1/query
>           - prefix: /prometheus/account/*/api/v1/query_range
>             replacement: /admin/prometheus/api/v1/query_range
>       services:
>         - name: promauthproxy
>           port: 8080
>       authPolicy:
>         context:
>           setPromAuthProxyHeader: "true"
> ```
> 
> I am trying to match the wildcard value, in my case a number, rewrite the path and set the ID in a header if the user passes auth. The above doesn't work. Unfortunately, I am not able to to rewrite a wildcard path. I've tried also setting a prefix condition like this: `/prometheus/account/*/` but to no avail. Is there any way for a Contour HttpProxy to rewrite a path `/some/account/*/path` to `/admin/path` where the wildcard is just a number (i.e. [0-9]+)?
> 
> I also tried changing my `promauthproxy` definition to:
> 
> ```yaml
> apiVersion: projectcontour.io/v1
> kind: HTTPProxy
> metadata:
>   name: promauthproxy
>   namespace: customer-monitoring
> spec:
>   routes:
>     - pathRewritePolicy:
>         replacePrefix:
>           - replacement: /admin/prometheus
>       services:
>         - name: promauthproxy
>           port: 8080
>       authPolicy:
>         context:
>           setPromAuthProxyHeader: "true"
> ```
> 
> But unfortunately the above doesn't work either. If I try to access
> 
> ```
> /prometheus/account/5/api/v1/query?query=<some_query>
> ```
> 
> then it gets rewritten to:
> 
> ```
> /admin/prometheus/5/api/v1/query?query=<some_query>
> ```
> 
> This is a showstopper for us and I have to switch back to using Nginx : ( Is there anything I can do in Contour to get rid of the `/5` in the above example?

**Comment by youngnick** [comment]
@jpeach did some checking of various regex program sizes in https://github.com/projectcontour/contour/issues/2043#issuecomment-565192280, reproduced here:

RE2 program sizes:

| regex | size |
|-------|-----|
| `.*foo.*'` | 35 |
| `foo.*[abcd]+ooo.*` | 43 |
| `x-envoy-(a\|b\|c)` | 15 |
| `x-envoy-(a\|b\|c)-[grep]{3}` | 28 |
| `a.*b.*c.*d.*e.*f.*g.*` | 109 |

I agree that we should probably set a larger regex max program size by default - although I would like @sunjayBhatia or @wrowe to comment on what having a larger regex max program size costs. Smells like a tradeoff to me.

We should also allow people to configure the max program size, but that will need some explanation of any tradeoffs involved.

I agree we should validate the syntax and ideally, check the program size. I think that we should also ensure that the way this is specified is via two keys, a find and a replace, rather than specifying the regex directly.

I think we should also make it so that if you are rewriting the path with this functionality, you cannot use includes, and vice versa.

**Comment by pavelnikolov** [comment]
In my particular case as well as another similar issues I saw even the default program size of 100 would work. We need to replace `path\/[0-9]{2,6}\/tail` with `path/tail`. Let's leave the current default and make it configurable. Also path rewriting is not supposed to brake any existing Contour installation since this is a new feature.

>We should also allow people to configure the max program size, but that will need some explanation of any tradeoffs involved.
💯 

> I think we should also make it so that if you are rewriting the path with this functionality, you cannot use includes, and vice versa.
What is the problem with rewriting path in includes? The routing is already done, no further routing decisions are to be made after the rewrite even if a route that matches the new path exists. That would simplify the task, right?

**Comment by skriss** [comment]
> I agree that we should probably set a larger regex max program size by default - although I would like @sunjayBhatia or @wrowe to comment on what having a larger regex max program size costs. Smells like a tradeoff to me.

For reference, it had been set to 1,048,576 from Contour versions 1.2 to 1.15, but we inadvertently then flipped it back to 100 when the xDS field was deprecated. See https://github.com/projectcontour/contour/pull/2241 for the original change that set it to ~1M. IMHO we should switch it back to 1,048,576 since the reversion to 100 was inadvertent.

**Comment by sunjayBhatia** [comment]
> larger regex max program size costs

larger program size pretty much means more memory footprint, since there are more instructions to evaluate, so in practice if you have some alerting on envoy memory usage you may see some increase, though it may not be obvious to an operator why

did a quick experiment with increasing the program size limit and Ingress path matching with some more expensive regexes just now to see if this can be quantified, but you have to get to significant scale (# of Ingress resources/routes and requests to evaluate) to see the impact, my simple test wasn't large enough to show anything substantial

right now though, something with a regex for 2 GUIDs for example will be rejected by Envoy (anchor added by Contour as a simple regex optimization)

```
gRPC config for type.googleapis.com/envoy.config.route.v3.RouteConfiguration rejected: regex '^/foo/[a-zA-Z0-9-]{0,16}/bar/[a-zA-Z0-9-]{0,16}' RE2 program size of 137 > max program size of 100 set for the error level threshold. Increase configured max program size if necessary.
```

**Comment by sunjayBhatia** [comment]
I think Dave's concerns in https://github.com/projectcontour/contour/pull/1612#issuecomment-537734352 are more around the routing/inclusions issues with HTTPProxy+regex matching, rather than path rewriting, since as @pavelnikolov points out, routing decisions are already done at the point path rewrites occur

the one gotcha may be that users have to have a good handle on what the path prefix will be actually be if rewriting on a child HTTPProxy (though if youre using includes you need to already in order to understanding how your request will be routed)

**Comment by sunjayBhatia** [comment]
to make max program size configurable, I would rather use RTDS rather than add flags to the Envoy bootstrap, its a bit more future-proof of a solution

if Envoy is updated before Contour and has bootstrap to reach out for Runtime configuration and it connects to a Contour that hasn't been updated with RTDS support, it will still start up but will log a few warnings:

```
[warning][config] [source/common/config/grpc_subscription_impl.cc:118] gRPC config: initial fetch timed out for type.googleapis.com/envoy.service.runtime.v3.Runtime
[info][runtime] [source/common/runtime/runtime_impl.cc:446] RTDS has finished initialization
...
[warning][config] [./source/common/config/grpc_stream.h:160] StreamRuntime gRPC config stream closed: 12, unknown service envoy.service.runtime.v3.RuntimeDiscoveryService
```

seems relatively acceptable to do this upgrade within one release since Envoy still starts up, but if we wanted to be conservative, we could introduce RTDS server-side support in the next Contour version (returning an empty set of Runtime layers e.g.) along with static configuration of regex program size, and offer configurability in the next version

**Comment by youngnick** [comment]
> the one gotcha may be that users have to have a good handle on what the path prefix will be actually be if rewriting on a child HTTPProxy (though if youre using includes you need to already in order to understanding how your request will be routed)

Thanks, @sunjayBhatia, this is what I was trying (poorly) to say, and I had conflated having globs in routing as well. Sorry for being unclear, @pavelnikolov.

With respect to using RTDS, I think it's better to be more conservative here, unless there's a strong tradeoff with having a large max program size.

**Comment by pavelnikolov** [comment]
What is an RTDS? 
By the way, take a look at this PR in the Kuma project - they also use Envoy behind the scenes and they have had the same project https://github.com/kumahq/kuma/pull/2139

**tl;dr** they use `uint32` max val for the max program size and log a warning when the size is greater than 1000, for example:
```
[2021-06-09 23:52:07.442][19][warning][misc] [source/common/common/regex.cc:100] regex '/static/[a-zA-Z0-9._-]{0,200}/[a-zA-Z0-9._-]{0,200}' RE2 program size of 2013 > max program size of 1000 set for the warn level threshold. Increase configured max program size if necessary.

```

**Comment by sunjayBhatia** [comment]
> What does RTDS mean?

Runtime discover service: https://www.envoyproxy.io/docs/envoy/latest/configuration/operations/runtime#config-runtime-rtds

**Comment by sunjayBhatia** [comment]
you can also monitor stats on regex program size which is nice, you don't have to just rely on memory usage: https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/matcher/v3/regex.proto.html?highlight=warn_level#type-matcher-v3-regexmatcher-googlere2

```
Envoy emits two stats for tracking the program size of regexes: the histogram re2.program_size, which records the program size, and the counter re2.exceeded_warn_level, which is incremented each time the program size exceeds the warn level threshold.
```

**Comment by pavelnikolov** [comment]
Now that the above PR is merged, what's left? Can you give me some guidance on where to look and I might try to work on a PR?

**Comment by sunjayBhatia** [comment]
I think the next steps on this are:
- add a "regex replacement" type and use it as a field in the `PathRewritePolicy` struct: https://github.com/projectcontour/contour/blob/c1094dbae2b501fdc633ea65b427c178d9f634b4/apis/projectcontour/v1/httpproxy.go#L844-L854
- add validation of the regex (probably ensure it compiles etc.) and the replacement (mindful of capture groups, etc.) in the `httpproxy_processor.go` file (e.g for prefix: https://github.com/projectcontour/contour/blob/c1094dbae2b501fdc633ea65b427c178d9f634b4/internal/dag/httpproxy_processor.go#L575)
- you can follow code paths for the existing `PathRewritePolicy`
  - e.g. where it is added to the dag structure for a route: https://github.com/projectcontour/contour/blob/c1094dbae2b501fdc633ea65b427c178d9f634b4/internal/dag/dag.go#L227
  - and then converted to Envoy config: https://github.com/projectcontour/contour/blob/c1094dbae2b501fdc633ea65b427c178d9f634b4/internal/envoy/v3/route.go#L267

**Comment by pavelnikolov** [comment]
I'm still interested in this feature. I'll give it a go soon.
